HPCC-24897 Fix esdl tool xsl parameter handling

asselitx opened the pull request:
Changes introduced in HPCC-22357 treat xsl parameters to the
EsdlDefinitionHelper as expressions, requiring strings to be quoted.
Institute correct string quoting in the esdl tool where it uses
the EsdlDefinitionHelper.

Add regression test cases to exercise the esdl tool's xsd, wsdl and cpp
commands.

Signed-off-by: Terrence Asselin <terrence.asselin@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-24897 Fix esdl tool xsl parameter handling

Changes introduced in HPCC-22357 treat xsl parameters to the
EsdlDefinitionHelper as expressions, requiring strings to be quoted.
Institute correct string quoting in the esdl tool where it uses
the EsdlDefinitionHelper.

Add regression test cases to exercise the esdl tool's xsd, wsdl and cpp
commands.

Signed-off-by: Terrence Asselin <terrence.asselin@lexisnexisrisk.com>

asselitx commented:
@timothyklemm and @afishbeck could you please review?

HPCCSmoketest commented:
Process of PR-14273, label: esdl-tool-param-fix-hpcc-24897 starts now.
The reason of this test is: Forced to re-test.
Commit ID: c86d45ecd4faca4ad50204b45e76387d4700ac0d
Estimated completion time is ~0.94 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-145.ca-central-1.compute.internal

HPCCSmoketest commented:
Process of PR-14273, label: esdl-tool-param-fix-hpcc-24897 starts now.
The reason of this test is: Did not build yet.
Commit ID: c86d45ecd4faca4ad50204b45e76387d4700ac0d
Estimated completion time is ~1.12 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-145.ca-central-1.compute.internal
Sha: c86d45ecd4faca4ad50204b45e76387d4700ac0d
Build: success
Milestone:Install hpccsystems-platform-community_7.12.5-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |51 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |42 sec  (00:00:42) |
|setup (roxie) |11 |11 |0 |23 sec  (00:00:23) |
|test (hthor) |934 |934 |0 |284 sec  (00:04:44) |
|test (thor) |845 |845 |0 |656 sec  (00:10:56) |
|test (roxie) |1013 |1013 |0 |352 sec  (00:05:52) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  633 sec (00:10:33) |  208 sec (00:03:28) |   20 sec (00:00:20) |   16 sec (00:00:16) | 1606 sec (00:26:46) |   17 sec (00:00:17) | 2509 sec (00:41:49) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: c86d45ecd4faca4ad50204b45e76387d4700ac0d
Build: success
Milestone:Install hpccsystems-platform-community_7.12.5-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |76 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |9 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |30 sec  (00:00:30) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |934 |934 |0 |652 sec  (00:10:52) |
|test (thor) |845 |845 |0 |1074 sec  (00:17:54) |
|test (roxie) |1013 |1013 |0 |770 sec  (00:12:50) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   14 sec (00:00:14) | 1186 sec (00:19:46) |  255 sec (00:04:15) |   23 sec (00:00:23) |   17 sec (00:00:17) | 2850 sec (00:47:30) |   17 sec (00:00:17) | 4362 sec (01:12:42) |

timothyklemm commented on esp/esdllib/esdl_def_helper.cpp:
```diff
@@ -249,9 +249,15 @@ void EsdlDefinitionHelper::toXSD( IEsdlDefObjectIterator &objs, StringBuffer &xs
         IProperties* params = *( parameters.getValue(xslId) );
-        const char *tns = (ns) ? ns :params->queryProp("tnsParam");
-
-        xml.appendf("<esxdl name=\"custom\" EsdlXslTypeId=\"%d\" xmlns:tns=\"%s\" ns_uri=\"%s\">", xslId, tns ? tns : "urn:unknown", tns ? tns : "urn:unknown");
+        StringBuffer tns((ns) ? ns :params->queryProp("tnsParam"));
+        // Since the xsl string params are quoted with ' we must remove
+        // it here before including it as the value of an attribute in
+        // the esxdl element
+        tns.stripChar('\'');
+        if( tns.isEmpty() )
+            tns.set("urn:unknown");
+
+        xml.appendf("<esxdl name=\"custom\" EsdlXslTypeId=\"%d\" xmlns:tns=\"%s\" ns_uri=\"%s\">", xslId, tns.str(), tns.str());
         this->toXML( objs, xml, version, opts, flags );
         xml.append("</esxdl>");
```
This block was pretty simple and might not have warranted a stand-alone function, but the increased complexity may now justify a bit of refactoring to avoid deviations between the two occurrences.

asselitx replied:
Created a class method to encapsulate the logic of determining which namespace to use and unquoting it if required.

timothyklemm commented on tools/esdlcmd/esdlcmd_core.cpp:
```diff
@@ -328,17 +328,20 @@ class Esdl2XSDCmd : public EsdlHelperConvertCmd
+        // All params are treated as expressions, so any strings must be quoted
+        // 1/0 are not equivalent to true/false, as 0 evaluates to true
+
+        params->setProp( "tnsParam", StringBuffer('\'').append(tns).append('\'') );
```
I would use VStringBuffer("'%s'", tns.str()) instead of StringBuffer('\'').append(tns).append('\'').

asselitx replied:
Done, wrapped in macro.

timothyklemm commented:
I can appreciate the appeal of using a real world example, but I think a minimalist example that clearly exercises the points of concern might be better. Does this example cover everything we care about today? Will it cover what we care about after the next ESDL revision and, if not, would we inject deviations from the real world into the existing test case or would we be expected to mimic a different real world example that may or may not be comprehensive?

Commit:
HPCC-24897 Make changes recommended in code review

 - Remove WsDfu test cases and use smaller targeted example service.
 - Improve string parameter quoting.
 - Factor out common code from two EsdlDefHelper methods.
 - Update test script with new test cases and make small readability
   and result statistics tracking fixes.

Signed-off-by: Terrence Asselin <terrence.asselin@lexisnexisrisk.com>

HPCCSmoketest commented:
Process of PR-14273, label: esdl-tool-param-fix-hpcc-24897 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 6bfcd28d199733cd0c9ac2e2659235ed05ba7a16
Estimated completion time is ~0.94 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-238.ca-central-1.compute.internal

HPCCSmoketest commented:
Process of PR-14273, label: esdl-tool-param-fix-hpcc-24897 starts now.
The reason of this test is: New commit.
Commit ID: 6bfcd28d199733cd0c9ac2e2659235ed05ba7a16
Estimated completion time is ~1.12 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-238.ca-central-1.compute.internal
Sha: 6bfcd28d199733cd0c9ac2e2659235ed05ba7a16
Build: success
Milestone:Install hpccsystems-platform-community_7.12.5-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |40 sec  (00:00:40) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |935 |935 |0 |277 sec  (00:04:37) |
|test (thor) |846 |846 |0 |715 sec  (00:11:55) |
|test (roxie) |1014 |1014 |0 |352 sec  (00:05:52) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  645 sec (00:10:45) |  209 sec (00:03:29) |   21 sec (00:00:21) |   16 sec (00:00:16) | 1655 sec (00:27:35) |   17 sec (00:00:17) | 2572 sec (00:42:52) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 6bfcd28d199733cd0c9ac2e2659235ed05ba7a16
Build: success
Milestone:Install hpccsystems-platform-community_7.12.5-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |76 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |9 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |27 sec  (00:00:27) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |935 |935 |0 |641 sec  (00:10:41) |
|test (thor) |846 |846 |0 |1122 sec  (00:18:42) |
|test (roxie) |1014 |1014 |0 |770 sec  (00:12:50) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   12 sec (00:00:12) | 1180 sec (00:19:40) |  250 sec (00:04:10) |   27 sec (00:00:27) |   17 sec (00:00:17) | 2885 sec (00:48:05) |   17 sec (00:00:17) | 4388 sec (01:13:08) |

timothyklemm commented on testing/esp/esdlcmd/README.rst:
```diff
@@ -0,0 +1,26 @@
+baselines are kept in the testing/esp/esdlcmd/baselines directory. Each entry in the baselines
+directory must match the name of Test Case it is for.
+
+For example, the Test Case named 'wsdl1' to generate the wsdl for WsDfu, has a file named
+
+    'wsdl1.wsdl'
+
+in the baselines directory. Likewise, the Test Case named
+
+    'cpp-dfu1'
+
+that generates c++ files for the WsDfu service has a folder named 'cpp-dfu1' in the baselines
+directory.
```
WsDfu is no longer the example.

asselitx replied:
Ok I'll fix those comments.

asselitx replied:
Ok I'll fix those comments.

timothyklemm commented on testing/esp/esdlcmd/baselines/wstest-cpp-installdir/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+cmake_minimum_required(VERSION 3.0)
```
Copyright and license info? Comment applies to all new files, with the possible exception of the readme.

asselitx replied:
These are auto-generated by the esdl tool, and are added here not as code but as baseline output for the test cases. If I add copyright comments then I'd need to take that into account when making the comparison against the test case output. I'd like to avoid that if I can but if the comments are required even this situation I'll do that.

timothyklemm commented on testing/esp/esdlcmd/baselines/wstest-cpp-installdir/WsTestServiceBase.cpp:
```diff
@@ -0,0 +1,527 @@
+extern "C" int onWsTestAllVersionReport(const char* CtxStr, const char* ReqStr, StringBuffer& RespStr)
+{
+    Owned<WsTestServiceBase> service = createWsTestServiceObj();
+    return service->AllVersionReport(CtxStr, ReqStr, RespStr);
```
It seems odd to return the 64 bit address of an allocated buffer as a 32 bit integer.

timothyklemm commented on testing/esp/esdlcmd/baselines/wstest-cpp-installdir/WsTestService.hpp:
```diff
@@ -0,0 +1,35 @@
+class WsTestService : public WsTestServiceBase
+{
+public:
+        virtual AllVersionReportResponse* AllVersionReport(EsdlContext* context, AllVersionReportRequest* request);
+    virtual MinVersionReportResponse* MinVersion(EsdlContext* context, MinVersionReportRequest* request);
+    virtual WsTestPingResponse* Ping(EsdlContext* context, WsTestPingRequest* request);
+    virtual VersionRangeReportResponse* VersionRangeReport(EsdlContext* context, VersionRangeReportRequest* request);
+    virtual int AllVersionReport(const char* CtxStr, const char* ReqStr, StringBuffer& RespStr)
+    {
+        return WsTestServiceBase::AllVersionReport(CtxStr, ReqStr, RespStr);
+    }
+    virtual int MinVersion(const char* CtxStr, const char* ReqStr, StringBuffer& RespStr)
+    {
+        return WsTestServiceBase::MinVersion(CtxStr, ReqStr, RespStr);
+    }
+    virtual int Ping(const char* CtxStr, const char* ReqStr, StringBuffer& RespStr)
+    {
+        return WsTestServiceBase::Ping(CtxStr, ReqStr, RespStr);
+    }
+    virtual int VersionRangeReport(const char* CtxStr, const char* ReqStr, StringBuffer& RespStr)
```
Can 'virtual' be replaced with 'override' where appropriate?

asselitx replied:
This is auto-generated code by the esdl tool. I didn't comb through the code in detail, but included the case mostly to verify the correct handling of the INSTALL_DIR parameter.

timothyklemm commented on testing/esp/esdlcmd/baselines/wstest-cpp-installdir/WsTestServiceBase.cpp:
```diff
@@ -0,0 +1,527 @@
+    }
+
+};
+int WsTestServiceBase::AllVersionReport(const char* CtxStr, const char* ReqStr, StringBuffer& RespStr)
+{
+    Owned<WsTestUnSerializer> UnSe = new WsTestUnSerializer();
+    Owned<EsdlContext> ctx = new EsdlContext();
+    if (CtxStr && *CtxStr)
+    {
+        Owned<IPropertyTree> ptree = createPTreeFromXMLString(CtxStr);
+        UnSe->unserialize(ctx.get(), ptree.get());
+    }
+    Owned<WsTestServiceBase> service = createWsTestServiceObj();
+    Owned<AllVersionReportRequest> req = new AllVersionReportRequest();
+    Owned<IPropertyTree> ptree;
+    ptree.setown(createPTreeFromXMLString(ReqStr));
+    UnSe->unserialize(req.get(), ptree.get());
+    Owned<AllVersionReportResponse> resp = service->AllVersionReport(ctx.get(), req.get());
+    RespStr = "<Response><Results><Result><Dataset name=\"AllVersionReportResponse\"><Row>";
+    UnSe->serialize(resp.get(), RespStr);
+    RespStr.append("</Row></Dataset></Result></Results></Response>");
+
+    return 0;
+}
```
If the service instance (line 453) overloaded a single method name (line 458), this method could be generalized as a template method to eliminate (currently) three additional nearly identical copies.

timothyklemm commented on testing/esp/esdlcmd/baselines/wstest-cpp-installdir/WsTestServiceBase.hpp:
```diff
@@ -0,0 +1,167 @@
+using namespace std;
+using namespace cppplugin;
+
+class EsdlContext : public CInterface, implements IInterface
```
Implementing IInterface is only necessary if you intend to treat the instance as an IInterface (which is not necessary for use with Owned). Is there a reason for inheriting both CInterface and IInterface in so many places?

asselitx replied:
There may be some opportunities to improve the auto-generated code, but I'm not going to address it as part of this ticket.

timothyklemm commented on testing/esp/esdlcmd/esdlcmd-test.py:
```diff
@@ -0,0 +1,368 @@
+            logging.error('Missing output for test %s', self.name)
+            return False
+
+        if (not filecmp.cmp(str(baseline), str(outName))):
```
I don't know python, so can't comment too much. One question, though, is whether equality is necessary or equivalence might be sufficient? In other words, is there an expectation that one input will yield the exact same ordering and whitespace every time?

asselitx replied:
I think that we should check for equality. Tony has mentioned that the wsdls and xsds generated should preserve order from the interface definitions.

Commit:
HPCC-24897 Address code review comments

 - Rename baselines directory and other references to 'key' to match other
   tests.
 - Update README to explain that key files are generated by the esdl tool, and
   explain how to update test cases.
 - Rename and simplify the QUOTED macro, and add an inline helper function
   for quoting params.

Signed-off-by: Terrence Asselin <terrence.asselin@lexisnexisrisk.com>

HPCCSmoketest commented:
Process of PR-14273, label: esdl-tool-param-fix-hpcc-24897 starts now.
The reason of this test is: Forced to re-test.
Commit ID: a66e2fa4d3733c284edf442fcbdc0215a5d0b96b
Estimated completion time is ~0.94 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-51.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-51.ca-central-1.compute.internal
Sha: a66e2fa4d3733c284edf442fcbdc0215a5d0b96b
Build: success
Milestone:Install hpccsystems-platform-community_7.12.11-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |146 |146 |0 |0 |0 |53 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |41 sec  (00:00:41) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |936 |936 |0 |274 sec  (00:04:34) |
|test (thor) |846 |846 |0 |651 sec  (00:10:51) |
|test (roxie) |1015 |1015 |0 |352 sec  (00:05:52) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  642 sec (00:10:42) |  211 sec (00:03:31) |   21 sec (00:00:21) |   16 sec (00:00:16) | 1601 sec (00:26:41) |   18 sec (00:00:18) | 2517 sec (00:41:57) |

asselitx commented:
@timothyklemm and @rpastrana could you please review this latest change?

rpastrana reviewed (approved):
@asselitx looks good. Skimmed over the tests and it ** seems ** fine to me. I left a couple of minor comments.
Also, are some of the classes in wstest-cpp-installdir generated? If so, does it make sense to generate at build time? Otherwise, do they need to be updated manually everytime the ECM files are updated?

timothyklemm reviewed (approved):
The readme reintroduced a reference to WsDfu, which is no longer the basis for the test case. But that seems trivial.
